test/resauce/core_test.clj

```diff
@@ -1,14 +1,216 @@
+     (juxt fs/file
+           fs/path
+           (memfn ^URI toURL)
+           identity
+           strip-file-prefix
+           str)
+
+     (string? f)
+     (juxt fs/file
+           fs/path
+           (comp (memfn ^URI toURL) str->uri)
+           str->uri
+           (if (prefixed-w-file? f) strip-file-prefix identity)
+           (if (prefixed-w-file? f) identity prefix-with-file)))
+   f))
+
+(defn some-file-path ^Path []
+  (->> (fs/list-dir (fs/home))
+       (filter fs/regular-file?)
+       (first)))
+
+(defn some-file-url ^URL []
+  ;; NB: Use some concrete resource that we know lives on the classpath as file.
+  (io/resource "resauce/core.clj"))
+
+(defn fs-files []
+  (mapcat ->file-reprs
+          [(some-file-path)
+           (some-file-url)]))
+
+(defn- ->dir-endings [f]
+  ((juxt identity #(str % "/")) (str f)))
+
+(defn some-fs-dirs []
+  ;; NB: Use some concrete resources that we know live on the classpath as dirs.
+  [(str (fs/home))
+   (str (io/as-file (io/resource "resauce")))])
+
+(defn correct-fs-dirs []
+  (->> (some-fs-dirs)
+       (mapcat ->dir-endings)
+       (mapcat ->file-reprs)
+       (distinct)))
+
+(defn incorrect-fs-dirs []
+  (->> (some-fs-dirs)
+       (mapcat ->dir-endings)
+       (map prefix-with-file)
+       (mapcat (juxt fs/file
+                     fs/path))
+       (distinct)))
+
+;;; JARs
+
+;; NB: Syntax is from `java.net.JarURLConnection`.
+(def jar-url-re #"jar:(?<url>[^!]+)!/(?<entry>.+)?")
+
+(defn- ->jar-file
+  "Yields 'file:/<...>.jar' string."
+  [^URL jar-url]
+  (when-some [[_ url _entry] (re-matches jar-url-re (str jar-url))]
+    url))
+
+(defn- ->jar-file-w-trailing-bang
+  "Yields 'file:/<...>.jar!' string."
+  [^URL jar-url]
+  (str (->jar-file jar-url) "!"))
+
+(defn- ->jar-root-dir
+  "Yields 'jar:file:/<...>.jar!/' string."
+  [^URL jar-url]
+  (str "jar:" (->jar-file jar-url) "!/"))
+
+(defn- ->jar-meta-inf-dir
+  "Yields 'jar:file:/<...>.jar!/META-INF' string."
+  [^URL jar-url]
+  (str "jar:" (->jar-file jar-url) "!/META-INF"))
+
+(defn some-jar-file-urls []
+  ;; NB: Use some concrete resources that we know live on the classpath in JARs.
+  [(io/resource "clojure/core.clj")
+   (io/resource "META-INF/leiningen/medley/medley/README.md")])
+
+(defn files-in-jars []
+  (mapcat (juxt str
+                identity
+                (comp str->uri str))
+          (some-jar-file-urls)))
+
+(defn- ->jar-file-reprs
+  "URL | URI | \"...\" -> [URL
+                           URI
+                           \"...\"]"
+  [f]
+  ((cond
+     (instance? URL f)
+     (juxt identity
+           (memfn ^URL toURI)
+           str)
+
+     (instance? URI f)
+     (juxt (memfn ^URI toURL)
+           identity
+           str)
+
+     (string? f)
+     (juxt io/as-url
+           str->uri
+           identity))
+   f))
+
+(defn some-jar-dir-urls []
+  ;; NB: Use some concrete resources that we know live on the classpath in JARs.
+  [(io/resource "clojure")
+   (io/resource "medley")])
```

Comment:
An already used `"clojure"` JAR dir is complemented by the `"medley"` dir, also JARed.